### DISCUSSION TITLE
How can I run a command when the container shuts down, to perform clean-up work?

### DISCUSSION THREAD

**Discussion by JaneSmith** [devcontainers/community#162] [Q&A]
Hello. As the title says, my goal is run a command when the dev container shuts down. I want to perform clean-up work, such as deleting the /node_modules/ directory in the project folder.

The `devcontainer.json` file has convenient `postCreateCommand` and `postStartCommand` options, but apparently nothing for shutdown. There is a `shutdownAction` defined, but seemingly unrelated. I'm essentially looking for a way to set up a `shutdownCommand`.

I attempted to do this by trapping EXIT in the postStartCommand, e.g.:

`"postStartCommand": "trap \"rm -rf ${containerWorkspaceFolder}/node_modules\" EXIT"`

Unfortunately, this didn't work as intended. Instead of running on shutdown, it runs at the end of the startup sequence, as it catches the startup sequence exiting.

What is the proper way to handle this?

**Comment by bamurtaugh** [comment]
Thanks for opening this discussion!

I believe the changes that could come with https://github.com/devcontainers/cli/issues/386 would help in this scenario (the CLI would need to handle this, since Docker wouldn't perform / manage shutdown).
